File: packages/core/src/verifier.ts
import { SignatureBaseError, buildSignatureBase, type StructuredFieldType } from './base.js'
import { importEd25519PublicKey, isEd25519, keyValidityAt, type Jwk } from './crypto.js'
import { verifyEd25519 } from './crypto.js'
import type { CacheResult, KeyResolver, NonceStore } from './keys.js'
import { DEFAULT_PROFILE, type Profile } from './profile.js'
import type { ReasonCode } from './reasons.js'
import { parseDictionary, parseItem } from './sfv/parse.js'
import type { BareItem, Dictionary, Item } from './sfv/types.js'
import { systemClock, type Clock, type NormalizedRequest, type Verdict } from './types.js'
import { verdictFor, type VerdictDetails } from './verdict.js'

export interface VerifierOptions {
  /** How a `keyid` becomes a public key. The only step permitted to do I/O. */
  readonly keys: KeyResolver
  /** Which draft revision's rules to apply. Recorded in every verdict. */
  readonly profile?: Profile
  readonly clock?: Clock
  /** How far into the future `created` may be, in seconds. */
  readonly clockSkewSec?: number
  /** Ceiling on `now - created`, regardless of what `expires` says. */
  readonly maxAgeSec?: number
  /**
   * When set, only these `Signature-Agent` origins are accepted.
   *
   * The strongest available answer to attacker-driven egress: nothing outside
   * this list is ever fetched.
   */
  readonly allowedOrigins?: readonly string[]
  /** Opt-in replay protection. Off by default — see ARCHITECTURE.md §8.4. */
  readonly replay?: NonceStore
  /**
   * Smallest nonce Badge will accept, in decoded bytes, when replay protection
   * is on.
   *
   * The reference implementation generates and requires exactly 64. Badge's
   * default is lower so a signer using a shorter but still unguessable nonce
   * interoperates, and high enough that the space cannot be enumerated and
   * pre-seeded into the replay store — which would turn replay protection into
   * a denial of service against the signer it is meant to protect. Set it to 64
   * to match the reference exactly.
   */
  readonly minNonceBytes?: number
  /**
   * Additions to the structured field type map, so `;sf` and `;key` can
   * canonicalize fields Badge does not know about.
   */
  readonly structuredFieldTypes?: Readonly<Record<string, StructuredFieldType>>
}

export interface Verifier {
  verify(request: NormalizedRequest): Promise<Verdict>
}

/** Accumulated as the pipeline learns things, so every exit reports what it knew. */
interface Facts {
  profile: string
  signatureAgent?: string
  keyid?: string
  label?: string
  created?: number
  expires?: number
  covered?: readonly string[]
  directoryUs?: number
  cache?: CacheResult
}

export function createVerifier(options: VerifierOptions): Verifier {
  const profile = options.profile ?? DEFAULT_PROFILE
  const clock = options.clock ?? systemClock
  const clockSkewSec = options.clockSkewSec ?? 5
  const maxAgeSec = options.maxAgeSec ?? 300
  const minNonceBytes = options.minNonceBytes ?? 16
  const allowedOrigins = options.allowedOrigins

  return {
    async verify(request: NormalizedRequest): Promise<Verdict> {
      const startedAt = performance.now()
      const facts: Facts = { profile: profile.id }
      const done = (reason: ReasonCode): Verdict =>
        verdictFor(reason, toDetails(facts, performance.now() - startedAt))

      try {
        // Step 1: the cheap presence test. Almost all production traffic exits
        // here, before any parsing, allocation, or I/O.
        const rawInput = request.header('signature-input')
        const rawAgent = request.header('signature-agent')
        if (rawInput === undefined) {
          return done(rawAgent === undefined ? 'no_signature_fields' : 'signature_input_malformed')
        }

        // Step 2: parse.
        let inputs: Dictionary
        try {
          inputs = parseDictionary(rawInput)
        } catch {
          return done('signature_input_malformed')
        }

        const rawSignature = request.header('signature')
        if (rawSignature === undefined) return done('signature_malformed')
        let signatures: Dictionary
        try {
          signatures = parseDictionary(rawSignature)
        } catch {
          return done('signature_malformed')
        }

        // Step 3: select the Web Bot Auth signature. Other signatures on the
        // message are ignored rather than treated as errors.
        const selected = selectWebBotAuth(inputs, profile.tag)
        if (selected === undefined) return done('no_web_bot_auth_tag')
        const { label, components, params, source } = selected
        facts.label = label
        facts.covered = components.map((c) => (c.value.type === 'string' ? c.value.value : '?'))

        // Step 4: preflight, cheapest and most attributable checks first.
        const keyid = stringParam(params, 'keyid')
        if (keyid === undefined) {
          if (profile.requireKeyid) return done('missing_keyid')
        } else {
          facts.keyid = keyid
        }

        const alg = stringParam(params, 'alg')
        if (alg !== undefined && !profile.algorithms.includes(alg)) {
          return done('unsupported_algorithm')
        }

        const created = integerParam(params, 'created')
        if (created === undefined) {
          if (profile.requireCreated) return done('missing_created')
        } else {
          facts.created = created
        }
        const expires = integerParam(params, 'expires')
        if (expires === undefined) {
          if (profile.requireExpires) return done('missing_expires')
        } else {
          facts.expires = expires
        }

        const missingComponent = checkCoveredComponents(components, request, profile)
        if (missingComponent !== undefined) return done(missingComponent)

        // Step 5: identity of the directory.
        let origin: string | undefined
        if (rawAgent !== undefined) {
          const parsed = parseSignatureAgent(rawAgent)
          if (parsed === undefined) return done('signature_agent_malformed')
          origin = parsed
          facts.signatureAgent = parsed
        } else if (profile.requireSignatureAgent) {
          return done('signature_agent_missing')
        }
        if (
          allowedOrigins !== undefined &&
          (origin === undefined || !allowedOrigins.includes(origin))
        ) {
          return done('signature_agent_not_allowed')
        }

        // Step 6: the window. Structural problems are reported before expiry so
        // a nonsense window reads as malformed rather than merely stale.
        const now = clock.now()
        if (created !== undefined && expires !== undefined) {
          if (expires - created > profile.maxWindowSec) return done('validity_window_too_long')
          if (expires < created) return done('validity_window_too_long')
        }
        if (created !== undefined && created > now + clockSkewSec) return done('created_in_future')
        if (expires !== undefined && now - clockSkewSec > expires) return done('signature_expired')
        if (created !== undefined && now - created > maxAgeSec) return done('signature_too_old')

        // Step 7: the signature bytes.
        const signatureBytes = signatureFor(signatures, label)
        if (signatureBytes === undefined) return done('signature_malformed')

        // Step 8: reconstruct the base.
        let base: string
        try {
          base = buildSignatureBase({
            request,
            components,
            signatureParamsSource: source,
            ...(options.structuredFieldTypes === undefined
              ? {}
              : { structuredFieldTypes: options.structuredFieldTypes }),
          })
        } catch (err) {
          if (err instanceof SignatureBaseError) return done(err.reason)
          throw err
        }

        // Step 9: resolve the key. The only I/O in the pipeline.
        if (keyid === undefined) return done('missing_keyid')
        const directoryStartedAt = performance.now()
        const resolution = await options.keys.resolve({ origin, keyid, now })
        facts.directoryUs = performance.now() - directoryStartedAt
        if (resolution.cache !== undefined) facts.cache = resolution.cache
        if (!resolution.ok) return done(resolution.reason)

        const jwk: Jwk = resolution.jwk
        if (!isEd25519(jwk)) return done('unsupported_algorithm')
        switch (keyValidityAt(jwk, now)) {
          case 'not-yet-valid':
            return done('key_not_yet_valid')
          case 'expired':
            return done('key_expired')
          case 'valid':
            break
        }

        // Step 10: verify.
        const publicKey = await importEd25519PublicKey(jwk)
        const verified = await verifyEd25519(
          publicKey,
          signatureBytes,
          new TextEncoder().encode(base),
        )
        if (!verified) return done('signature_invalid')

        // Step 11: replay, only when a store is configured.
        if (options.replay !== undefined) {
          const nonce = stringParam(params, 'nonce')
          if (nonce === undefined) return done('nonce_missing')
          if (decodedNonceBytes(nonce) < minNonceBytes) return done('nonce_invalid')
          let fresh: boolean
          try {
            fresh = await options.replay.checkAndRecord(
              nonce,
              replayRetainUntil(created, expires, clockSkewSec, maxAgeSec, now),
            )
          } catch {
            // A store outage must never read as a replay: that would deny
            // legitimate traffic the moment Redis hiccups.
            return done('nonce_store_unavailable')
          }
          if (!fresh) return done('replay_detected')
        }

        return done('ok')
      } catch {
        return done('internal_error')
      }
    },
  }
}

/**
 * The last instant this signature could still be accepted.
 *
 * A nonce must be remembered for exactly that long. Remembering it only until
 * `expires` leaves the signature replayable for the final `clockSkewSec`
 * seconds of its own acceptance window, because the expiry check allows
 * `now - clockSkewSec <= expires` — a hole that grows with any skew allowance
 * an operator raises.
 *
 * Acceptance needs both the expiry and the age check to hold, so the earlier of
 * the two bounds is the real deadline, and retaining past it would only waste
 * space in the store.
 */
function replayRetainUntil(
  created: number | undefined,
  expires: number | undefined,
  clockSkewSec: number,
  maxAgeSec: number,
  now: number,
): number {
  const byExpiry = (expires ?? now) + clockSkewSec
  if (created === undefined) return byExpiry
  return Math.min(byExpiry, created + maxAgeSec)
}

interface Selection {
  readonly label: string
  readonly components: readonly Item[]
  readonly params: ReadonlyMap<string, BareItem>
  readonly source: string
}

/**
 * Pick the first signature tagged for Web Bot Auth.
 *
 * The tag must be a String. A Token spelling `web-bot-auth` is a different
 * value in RFC 9651 and must not be accepted as one.
 */
function selectWebBotAuth(inputs: Dictionary, tag: string): Selection | undefined {
  for (const [label, entry] of inputs) {
    if (entry.value.kind !== 'inner-list') continue
    const tagParam = entry.value.params.get('tag')
    if (tagParam?.type !== 'string' || tagParam.value !== tag) continue
    return {
      label,
      components: entry.value.items,
      params: entry.value.params,
      source: entry.source,
    }
  }
  return undefined
}

function checkCoveredComponents(
  components: readonly Item[],
  request: NormalizedRequest,
  profile: Profile,
): ReasonCode | undefined {
  const names = new Set<string>()
  for (const component of components) {
    if (component.value.type !== 'string') return 'signature_input_malformed'
    names.add(component.value.value)
  }
  // Each group is satisfied by any one of its members, so a signature over
  // `@target-uri` satisfies the `@authority` requirement it subsumes.
  for (const group of profile.requiredComponents) {
    if (!group.some((name) => names.has(name))) return 'covered_components_insufficient'
  }
  for (const required of profile.requiredComponentsWhenPresent) {
    if (request.header(required) !== undefined && !names.has(required)) {
      return 'covered_components_insufficient'
    }
  }
  return undefined
}

/**
 * `Signature-Agent` is a structured field String holding an absolute `https:`
 * URI. Anything else — a bare token, an `http:` URI, a path — is malformed.
 */
function parseSignatureAgent(raw: string): string | undefined {
  let item: Item
  try {
    item = parseItem(raw)
  } catch {
    return undefined
  }
  if (item.value.type !== 'string') return undefined
  let url: URL
  try {
    url = new URL(item.value.value)
  } catch {
    return undefined
  }
  if (url.protocol !== 'https:') return undefined
  return url.origin
}

function signatureFor(signatures: Dictionary, label: string): Uint8Array | undefined {
  const entry = signatures.get(label)
  if (entry?.value.kind !== 'item') return undefined
  if (entry.value.value.type !== 'binary') return undefined
  return entry.value.value.value
}

/**
 * Decoded length of a base64 or base64url nonce, or 0 if it is not either.
 *
 * Computed from the encoded length rather than by decoding: the value is
 * attacker-controlled and there is no reason to allocate a buffer for it.
 */
function decodedNonceBytes(nonce: string): number {
  if (!/^[A-Za-z0-9+/_-]+={0,2}$/.test(nonce)) return 0
  const unpadded = nonce.replace(/=+$/, '').length
  return Math.floor((unpadded * 3) / 4)
}

function stringParam(params: ReadonlyMap<string, BareItem>, name: string): string | undefined {
  const value = params.get(name)
  return value?.type === 'string' ? value.value : undefined
}

function integerParam(params: ReadonlyMap<string, BareItem>, name: string): number | undefined {
  const value = params.get(name)
  return value?.type === 'integer' ? value.value : undefined
}

function toDetails(facts: Facts, elapsedMs: number): VerdictDetails {
  return {
    profile: facts.profile,
    signatureAgent: facts.signatureAgent,
    keyid: facts.keyid,
    label: facts.label,
    created: facts.created,
    expires: facts.expires,
    covered: facts.covered,
    timing: {
      totalUs: Math.round(elapsedMs * 1000),
      ...(facts.directoryUs === undefined
        ? {}
        : { directoryUs: Math.round(facts.directoryUs * 1000) }),
      ...(facts.cache === undefined ? {} : { cache: facts.cache }),
    },
  }
}

File: sidecar/wba/verifier.go
package wba

import (
	"context"
	"crypto/ed25519"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/ndunl075/badge/sidecar/sfv"
)

// Verdict is the verifier's answer. It always carries a reason, including on
// success.
type Verdict struct {
	Status         Status   `json:"status"`
	Class          Class    `json:"class"`
	Reason         Reason   `json:"reason"`
	Profile        string   `json:"profile"`
	SignatureAgent string   `json:"signature_agent,omitempty"`
	Keyid          string   `json:"keyid,omitempty"`
	Label          string   `json:"label,omitempty"`
	Created        int64    `json:"created,omitempty"`
	Expires        int64    `json:"expires,omitempty"`
	Covered        []string `json:"covered,omitempty"`
	Cache          string   `json:"cache,omitempty"`
	TotalUs        int64    `json:"total_us"`
	DirectoryUs    int64    `json:"directory_us,omitempty"`
}

// Clock is injected so tests can move time without touching the system clock.
type Clock interface{ Now() int64 }

type systemClock struct{}

func (systemClock) Now() int64 { return time.Now().Unix() }

// SystemClock reads the real time.
var SystemClock Clock = systemClock{}

// FixedClock is a clock a test can move.
type FixedClock struct{ Seconds int64 }

func (c *FixedClock) Now() int64 { return c.Seconds }

// KeyRequest asks a resolver for one key.
type KeyRequest struct {
	Origin string
	Keyid  string
	Now    int64
}

// Resolution is a resolver's answer.
type Resolution struct {
	OK     bool
	Key    JWK
	Reason Reason
	Cache  string
}

// KeyResolver finds the public key a keyid refers to.
//
// This is the only part of verification that may do I/O, and it is an interface
// precisely so the network implementation — with its caches, timeouts and SSRF
// guard — stays out of the verifier.
type KeyResolver interface {
	Resolve(ctx context.Context, req KeyRequest) Resolution
}

// NonceStore is an atomic check-and-record for replay protection. It must be
// atomic across the whole enforcement boundary; a per-process store behind more
// than one replica is theatre.
type NonceStore interface {
	CheckAndRecord(ctx context.Context, nonce string, expiresAt int64) (fresh bool, err error)
}

// StaticKeys resolves from a fixed set, with no network access at all.
type StaticKeys map[string][]JWK

func (s StaticKeys) Resolve(_ context.Context, req KeyRequest) Resolution {
	for _, origin := range []string{req.Origin, "*"} {
		if origin == "" {
			continue
		}
		for _, key := range s[origin] {
			thumb, err := Thumbprint(key)
			if err != nil {
				continue
			}
			if thumb == req.Keyid {
				return Resolution{OK: true, Key: key, Cache: "hit"}
			}
		}
	}
	return Resolution{Reason: ReasonKeyNotFound, Cache: "hit"}
}

// Verifier applies one profile's rules to incoming requests.
type Verifier struct {
	Keys                 KeyResolver
	Profile              Profile
	Clock                Clock
	ClockSkewSec         int64
	MaxAgeSec            int64
	MinNonceBytes        int
	AllowedOrigins       []string
	Replay               NonceStore
	StructuredFieldTypes map[string]StructuredFieldType
}

// NewVerifier fills in the defaults.
func NewVerifier(keys KeyResolver) *Verifier {
	return &Verifier{
		Keys:          keys,
		Profile:       DefaultProfile,
		Clock:         SystemClock,
		ClockSkewSec:  5,
		MaxAgeSec:     300,
		MinNonceBytes: 16,
	}
}

type facts struct {
	profile     string
	agent       string
	keyid       string
	label       string
	created     int64
	expires     int64
	covered     []string
	cache       string
	directoryUs int64
}

func (f facts) verdict(reason Reason, elapsed time.Duration) Verdict {
	status, class := reason.Info()
	return Verdict{
		Status: status, Class: class, Reason: reason, Profile: f.profile,
		SignatureAgent: f.agent, Keyid: f.keyid, Label: f.label,
		Created: f.created, Expires: f.expires, Covered: f.covered,
		Cache: f.cache, TotalUs: elapsed.Microseconds(), DirectoryUs: f.directoryUs,
	}
}

var base64ish = regexp.MustCompile(`^[A-Za-z0-9+/_-]+={0,2}$`)

// Verify runs the eleven steps of the verification pipeline.
func (v *Verifier) Verify(ctx context.Context, req *Request) Verdict {
	started := time.Now()
	f := facts{profile: v.Profile.ID}
	done := func(reason Reason) Verdict { return f.verdict(reason, time.Since(started)) }

	// Step 1: the cheap presence test. Almost all production traffic exits here,
	// before any parsing, allocation or I/O.
	rawInput, hasInput := req.Header("signature-input")
	rawAgent, hasAgent := req.Header("signature-agent")
	if !hasInput {
		if !hasAgent {
			return done(ReasonNoSignatureFields)
		}
		return done(ReasonSignatureInputMalformed)
	}

	// Step 2: parse.
	inputs, err := sfv.ParseDictionary(rawInput)
	if err != nil {
		return done(ReasonSignatureInputMalformed)
	}
	rawSignature, hasSignature := req.Header("signature")
	if !hasSignature {
		return done(ReasonSignatureMalformed)
	}
	signatures, err := sfv.ParseDictionary(rawSignature)
	if err != nil {
		return done(ReasonSignatureMalformed)
	}

	// Step 3: select the Web Bot Auth signature; ignore any others.
	label, selected, ok := selectWebBotAuth(inputs, v.Profile.Tag)
	if !ok {
		return done(ReasonNoWebBotAuthTag)
	}
	f.label = label
	params := selected.Member.InnerList.Params
	components := selected.Member.InnerList.Items
	f.covered = make([]string, 0, len(components))
	for _, c := range components {
		if c.Value.Kind == sfv.KindString {
			f.covered = append(f.covered, c.Value.Str)
		} else {
			f.covered = append(f.covered, "?")
		}
	}

	// Step 4: preflight, cheapest and most attributable checks first.
	keyid, hasKeyid := params.GetString("keyid")
	if !hasKeyid {
		if v.Profile.RequireKeyid {
			return done(ReasonMissingKeyid)
		}
	} else {
		f.keyid = keyid
	}
	if alg, ok := params.GetString("alg"); ok && !contains(v.Profile.Algorithms, alg) {
		return done(ReasonUnsupportedAlgorithm)
	}
	created, hasCreated := params.GetInteger("created")
	if !hasCreated {
		if v.Profile.RequireCreated {
			return done(ReasonMissingCreated)
		}
	} else {
		f.created = created
	}
	expires, hasExpires := params.GetInteger("expires")
	if !hasExpires {
		if v.Profile.RequireExpires {
			return done(ReasonMissingExpires)
		}
	} else {
		f.expires = expires
	}

	if reason, bad := checkCovered(components, req, v.Profile); bad {
		return done(reason)
	}

	// Step 5: identity of the directory.
	origin := ""
	if hasAgent {
		parsed, ok := parseSignatureAgent(rawAgent)
		if !ok {
			return done(ReasonSignatureAgentMalformed)
		}
		origin = parsed
		f.agent = parsed
	} else if v.Profile.RequireSignatureAgent {
		return done(ReasonSignatureAgentMissing)
	}
	if v.AllowedOrigins != nil && (origin == "" || !contains(v.AllowedOrigins, origin)) {
		return done(ReasonSignatureAgentNotAllowed)
	}

	// Step 6: the window. Structural problems are reported before expiry, so a
	// nonsense window reads as malformed rather than merely stale.
	now := v.Clock.Now()
	if hasCreated && hasExpires {
		if expires-created > v.Profile.MaxWindowSec || expires < created {
			return done(ReasonValidityWindowTooLong)
		}
	}
	if hasCreated && created > now+v.ClockSkewSec {
		return done(ReasonCreatedInFuture)
	}
	if hasExpires && now-v.ClockSkewSec > expires {
		return done(ReasonSignatureExpired)
	}
	if hasCreated && now-created > v.MaxAgeSec {
		return done(ReasonSignatureTooOld)
	}

	// Step 7: the signature bytes.
	signatureBytes, ok := signatureFor(signatures, label)
	if !ok {
		return done(ReasonSignatureMalformed)
	}

	// Step 8: reconstruct the base.
	base, err := BuildBase(req, components, selected.Source, v.StructuredFieldTypes)
	if err != nil {
		if be, ok := err.(*BaseError); ok {
			return done(be.Reason)
		}
		return done(ReasonInternalError)
	}

	// Step 9: resolve the key. The only I/O in the pipeline.
	if !hasKeyid {
		return done(ReasonMissingKeyid)
	}
	directoryStarted := time.Now()
	resolution := v.Keys.Resolve(ctx, KeyRequest{Origin: origin, Keyid: keyid, Now: now})
	f.directoryUs = time.Since(directoryStarted).Microseconds()
	f.cache = resolution.Cache
	if !resolution.OK {
		return done(resolution.Reason)
	}
	if !IsEd25519(resolution.Key) {
		return done(ReasonUnsupportedAlgorithm)
	}
	switch KeyValidityAt(resolution.Key, now) {
	case ValidityNotYetValid:
		return done(ReasonKeyNotYetValid)
	case ValidityExpired:
		return done(ReasonKeyExpired)
	}

	// Step 10: verify.
	pub, err := PublicKey(resolution.Key)
	if err != nil {
		return done(ReasonUnsupportedAlgorithm)
	}
	if !ed25519.Verify(pub, []byte(base), signatureBytes) {
		return done(ReasonSignatureInvalid)
	}

	// Step 11: replay, only when a store is configured.
	if v.Replay != nil {
		nonce, ok := params.GetString("nonce")
		if !ok {
			return done(ReasonNonceMissing)
		}
		if decodedNonceBytes(nonce) < v.MinNonceBytes {
			return done(ReasonNonceInvalid)
		}
		fresh, err := v.Replay.CheckAndRecord(ctx, nonce, v.replayRetainUntil(hasCreated, created, hasExpires, expires, now))
		if err != nil {
			// A store outage must never read as a replay: that would deny
			// legitimate traffic the moment the store hiccups.
			return done(ReasonNonceStoreUnavailable)
		}
		if !fresh {
			return done(ReasonReplayDetected)
		}
	}

	return done(ReasonOK)
}

// replayRetainUntil is the last instant this signature could still be accepted.
//
// Remembering a nonce only until expires would leave the signature replayable
// for the final ClockSkewSec seconds of its own acceptance window. Acceptance
// needs both the expiry and the age check to hold, so the earlier bound is the
// real deadline.
func (v *Verifier) replayRetainUntil(hasCreated bool, created int64, hasExpires bool, expires, now int64) int64 {
	byExpiry := now + v.ClockSkewSec
	if hasExpires {
		byExpiry = expires + v.ClockSkewSec
	}
	if !hasCreated {
		return byExpiry
	}
	byAge := created + v.MaxAgeSec
	if byAge < byExpiry {
		return byAge
	}
	return byExpiry
}

// decodedNonceBytes returns the decoded length of a base64 or base64url nonce,
// or 0 if it is neither. Computed from the encoded length rather than by
// decoding: the value is attacker-controlled and there is no reason to allocate.
func decodedNonceBytes(nonce string) int {
	if !base64ish.MatchString(nonce) {
		return 0
	}
	unpadded := strings.TrimRight(nonce, "=")
	return len(unpadded) * 3 / 4
}

// selectWebBotAuth picks the first signature tagged for Web Bot Auth. The tag
// must be a String; a Token spelling web-bot-auth is a different RFC 9651 value.
func selectWebBotAuth(inputs *sfv.Dictionary, tag string) (string, sfv.DictEntry, bool) {
	for _, label := range inputs.Keys() {
		entry, _ := inputs.Get(label)
		if !entry.Member.IsInnerList {
			continue
		}
		value, ok := entry.Member.InnerList.Params.GetString("tag")
		if !ok || value != tag {
			continue
		}
		return label, entry, true
	}
	return "", sfv.DictEntry{}, false
}

func checkCovered(components []sfv.Item, req *Request, profile Profile) (Reason, bool) {
	names := map[string]bool{}
	for _, component := range components {
		if component.Value.Kind != sfv.KindString {
			return ReasonSignatureInputMalformed, true
		}
		names[component.Value.Str] = true
	}
	// Each group is satisfied by any one of its members, so a signature over
	// @target-uri satisfies the @authority requirement it subsumes.
	for _, group := range profile.RequiredComponents {
		satisfied := false
		for _, name := range group {
			if names[name] {
				satisfied = true
				break
			}
		}
		if !satisfied {
			return ReasonCoveredComponentsInsufficient, true
		}
	}
	for _, required := range profile.RequiredComponentsWhenPresent {
		if _, present := req.Header(required); present && !names[required] {
			return ReasonCoveredComponentsInsufficient, true
		}
	}
	return "", false
}

// parseSignatureAgent reads the structured field String holding an absolute
// https URI. Anything else — a bare token, an http URI, a path — is malformed.
func parseSignatureAgent(raw string) (string, bool) {
	item, err := sfv.ParseItem(raw)
	if err != nil || item.Value.Kind != sfv.KindString {
		return "", false
	}
	parsed, err := url.Parse(item.Value.Str)
	if err != nil || parsed.Scheme != "https" || parsed.Host == "" {
		return "", false
	}
	return parsed.Scheme + "://" + parsed.Host, true
}

func signatureFor(signatures *sfv.Dictionary, label string) ([]byte, bool) {
	entry, ok := signatures.Get(label)
	if !ok || entry.Member.IsInnerList {
		return nil, false
	}
	if entry.Member.Item.Value.Kind != sfv.KindBinary {
		return nil, false
	}
	return entry.Member.Item.Value.Bin, true
}

func contains(list []string, want string) bool {
	for _, v := range list {
		if v == want {
			return true
		}
	}
	return false
}
